# Analysis without Model

In this notebook I will analyse the data of the musgrooms without using anytime of model, this will also help me to understand the dat I am working on. I will be expecting to be able to determine if the mushrooms are Edible or Poisonous only using the characteristics that are given to us.

In [24]:
import pandas as pd
import pandas as pd
import matplotlib as plt
import numpy as np

In [25]:
mushroom = pd.read_csv('../data/agaricus-lepiota.data')

In [26]:
mushroom.columns

Index(['p', 'x', 's', 'n', 't', 'p.1', 'f', 'c', 'n.1', 'k', 'e', 'e.1', 's.1',
       's.2', 'w', 'w.1', 'p.2', 'w.2', 'o', 'p.3', 'k.1', 's.3', 'u'],
      dtype='object')

In [27]:
mushroom.rename(columns = {'p':'classes','x':'cap-shape','s':'cap-surface','n':'cap-color',
                           't':'bruises','p.1':'odor','f':'gill-attachment','c':'gill-spacing',
                           'n.1':'gill-size','k':'gill-color','e':'stalk-shape','e.1':'stalk-root',
                           's.1':'stalk-surface-above-ring','s.2':'stalk-surface-below-ring',
                           'w':'stalk-color-above-ring','w.1':'stalk-color-below-ring',
                           'p.2':'veil-type','w.2':'veil-color','o':'ring-number','p.3':'ring-type',
                           'k.1':'spore-print-color','s.3':'population','u':'habitat'},inplace = True)

In [28]:
mushroom.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8123 entries, 0 to 8122
Data columns (total 23 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   classes                   8123 non-null   object
 1   cap-shape                 8123 non-null   object
 2   cap-surface               8123 non-null   object
 3   cap-color                 8123 non-null   object
 4   bruises                   8123 non-null   object
 5   odor                      8123 non-null   object
 6   gill-attachment           8123 non-null   object
 7   gill-spacing              8123 non-null   object
 8   gill-size                 8123 non-null   object
 9   gill-color                8123 non-null   object
 10  stalk-shape               8123 non-null   object
 11  stalk-root                8123 non-null   object
 12  stalk-surface-above-ring  8123 non-null   object
 13  stalk-surface-below-ring  8123 non-null   object
 14  stalk-color-above-ring  

From what I read on Mushrooms the characteristics that determine if a mushroom is edible or not are:

1. Cap color
2. Bruises
3. Odor
4. Gill color
5. Stalk surface Below ring
6. Spore print color
7. Ring number
8. Population
9. Habitat

I am going to use this to create a subset dataframe in which I will look at how this corresponds to the Mushroon being either Edible or Poisonous.

In [29]:
determiners = mushroom[['classes','cap-color','bruises','odor','gill-color',
                        'stalk-surface-below-ring','ring-number','spore-print-color',
                        'population','habitat']]

In [30]:
edible_mushroom = determiners[determiners.classes == "e"]

In [31]:
poisonous_mushroom = determiners[determiners.classes == "p"]

In [32]:
determiners.groupby('classes').count()

cap-color  bruises  odor  gill-color  stalk-surface-below-ring  \
classes                                                                   
e             4208     4208  4208        4208                      4208   
p             3915     3915  3915        3915                      3915   

         ring-number  spore-print-color  population  habitat  
classes                                                       
e               4208               4208        4208     4208  
p               3915               3915        3915     3915

From the count, we are able to see that in our data we have 4208 Mushrooms that are edible and 3915 which are Poisonous.

In [33]:
determiners.groupby('classes')['cap-color'].value_counts()

classes  cap-color
e        n            1264
         g            1032
         w             720
         e             624
         y             400
         p              56
         b              48
         c              32
         r              16
         u              16
p        n            1019
         e             876
         g             808
         y             672
         w             320
         b             120
         p              88
         c              12
Name: cap-color, dtype: int64

### Cap color

keys: [ brown=n, buff=b, cinnamon=c, gray=g, green=r, pink=p, purple=u, red=e, white=w, yellow=y ]

For Edible Mushrooms we have about 1264 mushrooms with a brown cap, 1032 with a grey cap, 720 with a white cap, 624 with a red cap, 400 with a yellow cap, 56 with a pink cap, 48 with a buff cap, 32 with a cinnamon cap, 16 with a green cap and 16 with a purple cap

For Poisonous Mushroom we have about 1019 with a brown cap, 876 with a red cap, 808 with a grey cap, 672 with a yellow cap, 320 with a white cap, 120 with a buff cap, 88 with a pink cap and 12 with a cinnamon cap.

In [34]:
 determiners.groupby('classes')['bruises'].value_counts()

classes  bruises
e        t          2752
         f          1456
p        f          3292
         t           623
Name: bruises, dtype: int64

### Bruises

keys: [bruises=t, no=f ]

For Edible Mushrooms we have about 2752 that have bruises and 1456 that do not have bruises.

For Poisonous Mushroom we have about 3292 that have no bruises and 623 with bruises.

In [35]:
determiners.groupby('classes')['odor'].value_counts()

classes  odor
e        n       3408
         a        400
         l        400
p        f       2160
         s        576
         y        576
         p        255
         c        192
         n        120
         m         36
Name: odor, dtype: int64

### Odor

keys: [ almond=a, anise=l, creosote=c, fishy=y, foul=f, musty=m, none=n, pungent=p, spicy=s ]

For Edible Mushroom we have about 3400 mushrooms with no odor, 400 with an almond odor and 400 with an anise odor.

For poisonous we have about 2160 mushrooms with a foul odor, 576 with a spicy odor, 576 with a fishy odor, 255 with a pungent odor, 192 with a creosote, 120 with no odor and  36 with a musty odor.



In [36]:
determiners.groupby('classes')['gill-color'].value_counts()

classes  gill-color
e        w              956
         n              936
         p              852
         u              444
         k              344
         g              248
         h              204
         e               96
         o               64
         y               64
p        b             1728
         p              640
         h              528
         g              504
         w              246
         n              112
         k               63
         u               48
         r               24
         y               22
Name: gill-color, dtype: int64

### Gill Color

Keys: [ black=k, brown=n, buff=b, chocolate=h, gray=g, green=r, orange=o, pink=p, purple=u, red=e, white=w, yellow=y ]

For Edible Mushroom, we have about  956 mushrooms with a white gill color, 936 with a brown gill color, 852 with a pink gill color,  444 with a purple gill color,  344 with a black gill color,  248 a grey gill color, 204 with a chocolate gill color,  96 with a red gill color,  64 with an orange gill color and 64 a yellow gill color.

For Poisonous Mushroom, we have about 1728 mushrooms with a buff gill color, 640 with a pink gill color, 528 with a chocolate gill color, 504 with a grey gill color, 246 with a white gill color, 112 with a brown gill color, 63 with a black gill color, 48 with a purple gill color, 24 with a green gill color and 22 with a yellow gill color.

In [37]:
determiners.groupby('classes')['stalk-surface-below-ring'].value_counts()

classes  stalk-surface-below-ring
e        s                           3400
         f                            456
         y                            208
         k                            144
p        k                           2160
         s                           1535
         f                            144
         y                             76
Name: stalk-surface-below-ring, dtype: int64

### Stalk surface Below ring

keys: [ fibrous=f, scaly=y, silky=k, smooth=s ]

With Edible Mushroom we have about 3400 mushrooms which have a smooth Surface, 456 with a fibrous surface, 208 with a scaly surfaceand 144 with a silky surface.

For Poisonous Mushroom we have about 2160 mushrooms that have a silky surface, 1535 with a smooth surface, 144 with a fibrous surface,


In [38]:
determiners.groupby('classes')['ring-number'].value_counts()

classes  ring-number
e        o              3680
         t               528
p        o              3807
         t                72
         n                36
Name: ring-number, dtype: int64

### Number of rings.

Keys: [ 	none=n, one=o, two=t ]

For Edible Mushroom, we have about 3680 with one  ring, 528 with two rings.

For Poisonous Mushroom, we have about 3807 with one ring, 72 with two rings and 36 with no rings at all.

In [39]:
determiners.groupby('classes')['spore-print-color'].value_counts()

classes  spore-print-color
e        n                    1744
         k                    1648
         w                     576
         b                      48
         h                      48
         o                      48
         u                      48
         y                      48
p        w                    1812
         h                    1584
         n                     224
         k                     223
         r                      72
Name: spore-print-color, dtype: int64

### Spore print color.

Keys: [ black=k, brown=n, buff=b, chocolate=h, green=r, orange=o, purple=u, white=w, yellow=y ]

For Edible Mushroom, we have about 1744 mushrooms with a brown spore print color, 1648 with a black spore print color, 576 with a white spore print color, 48 with a buff spore print color, 48 with a chocolate spore print color, 48 with an orange spore print color and 48 with a yellow spore print color.

For Poisonous Mushroom, we have about 1812 with a white spore print color, 1584 with a chocolate spore print color, 224 with a brown spore print color, 223 with a black spore print color and 72 with a green spore print color,

In [40]:
determiners.groupby('classes')['population'].value_counts()

classes  population
e        v             1192
         y             1064
         s              880
         n              400
         a              384
         c              288
p        v             2848
         y              648
         s              367
         c               52
Name: population, dtype: int64

### Population

Keys: [ abundant=a, clustered=c, numerous=n, scattered=s, several=v, solitary=y ]

For Edible Mushroom, we have about 1192 mushrooms that are several in population, 1094 which are in solidarity, 880 which are scattered,400 which are numerous,384 that are abundant and 288 that are clustered.

For Poisonous Mushroom, We have about 2848 mushrooms that are several in a population, 648 which are in solidarity, 367 that are scattered and 52 that are scattered.

In [41]:
determiners.groupby('classes')['habitat'].value_counts()

classes  habitat
e        d          1880
         g          1408
         m           256
         l           240
         w           192
         p           136
         u            96
p        d          1268
         p          1008
         g           740
         l           592
         u           271
         m            36
Name: habitat, dtype: int64

### Habitat 

Keys: [ grasses=g, leaves=l, meadows=m, paths=p, urban=u, waste=w, woods=d ]

For Edible Mushrooms, we have about 1880 which can be found in woods, 1408 on grasses, 256 in meadows, 240 on leaves, 192 in waste and 92 in urban areas.

For Poisonous Mushroom, we have about 1268 mushrooms that can be found in woods, 1008 on paths, 740 on grasses, 592 on leaves, 271 in urban areas and 36 in meadows.

# Analysis using a Model

In [42]:
from sklearn.manifold import TSNE
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.pipeline import Pipeline
from sklearn.metrics import confusion_matrix, classification_report

In [43]:
mushroom = pd.get_dummies(mushroom)
mushroom.head(0)

Empty DataFrame
Columns: [classes_e, classes_p, cap-shape_b, cap-shape_c, cap-shape_f, cap-shape_k, cap-shape_s, cap-shape_x, cap-surface_f, cap-surface_g, cap-surface_s, cap-surface_y, cap-color_b, cap-color_c, cap-color_e, cap-color_g, cap-color_n, cap-color_p, cap-color_r, cap-color_u, cap-color_w, cap-color_y, bruises_f, bruises_t, odor_a, odor_c, odor_f, odor_l, odor_m, odor_n, odor_p, odor_s, odor_y, gill-attachment_a, gill-attachment_f, gill-spacing_c, gill-spacing_w, gill-size_b, gill-size_n, gill-color_b, gill-color_e, gill-color_g, gill-color_h, gill-color_k, gill-color_n, gill-color_o, gill-color_p, gill-color_r, gill-color_u, gill-color_w, gill-color_y, stalk-shape_e, stalk-shape_t, stalk-root_?, stalk-root_b, stalk-root_c, stalk-root_e, stalk-root_r, stalk-surface-above-ring_f, stalk-surface-above-ring_k, stalk-surface-above-ring_s, stalk-surface-above-ring_y, stalk-surface-below-ring_f, stalk-surface-below-ring_k, stalk-surface-below-ring_s, stalk-surface-below-ring_y, stalk-color-above-ring_b, stalk-color-above-ring_c, stalk-color-above-ring_e, stalk-color-above-ring_g, stalk-color-above-ring_n, stalk-color-above-ring_o, stalk-color-above-ring_p, stalk-color-above-ring_w, stalk-color-above-ring_y, stalk-color-below-ring_b, stalk-color-below-ring_c, stalk-color-below-ring_e, stalk-color-below-ring_g, stalk-color-below-ring_n, stalk-color-below-ring_o, stalk-color-below-ring_p, stalk-color-below-ring_w, stalk-color-below-ring_y, veil-type_p, veil-color_n, veil-color_o, veil-color_w, veil-color_y, ring-number_n, ring-number_o, ring-number_t, ring-type_e, ring-type_f, ring-type_l, ring-type_n, ring-type_p, spore-print-color_b, spore-print-color_h, spore-print-color_k, ...]
Index: []

[0 rows x 119 columns]

In [44]:
LABELS = ['classes_e', 'classes_p']
FEATURES = [a  for a in mushroom.columns if a not in LABELS ]
y = mushroom[LABELS[0]]
x= mushroom[FEATURES]

In [45]:
#create model
model = TSNE(learning_rate = 100)

#fit model
transformed = model.fit_transform(x.values)
xs = transformed[:,0]
ys = transformed[:,1]
df_trans = pd.DataFrame({'xs':xs, 'ys':ys})

#create plots
plt.scatter(df_trans.loc[y==0]['xs'], df_trans.loc[y ==0]['ys'], c= 'tab:green')
plt.scatter(df_trans.loc[y ==1]['xs'], df_trans.loc[y ==1]['ys'], c= 'tab:blue')
plt.legend(loc ='lower left', labels = ['p', 'e'])

plt.show()

AttributeError: module 'matplotlib' has no attribute 'scatter'

In [ ]:
#create parameter grid for values of k
parameters = {'knn__n_neighbors': np.arange(1,10)}

#instatiate pipeline with KNNClassifier: pl
pl = Pipeline([('knn', KNeighborsClassifier())])

#split data into test and training data
X_train, X_test, y_train, y_test = train_test_split(x, y, test_size = 0.3, random_state = 42)

#instatiate GridsearchCV:cv
cv = GridSearchCV(pl, param_grid= parameters, cv = 3)

#fit model to training data
cv.fit(X_train, y_train)

#predict test data: y_pred
y_pred = cv.predict(X_test)

In [ ]:
#print performance metrics
print (cv.best_params_)
print(classification_report(y_test, y_pred))

In [ ]:
import graphviz
from sklearn import tree
from sklearn.tree import export_graphviz

#export a graphic representation of tree to file

dot_data =export_graphviz(tree, out_file = None, feature_names =x.columns, class_names = ['edible', 'poisonous'])
graph = graphviz.Source(dot_data)
graph